In recent Heimdal Kerberos GSS_C_NT_HOSTBASED_SERVICE is a macro
--- gsasl-1.8.1/lib/m4/gss-extra.m4.orig	2010-11-01 20:22:42.000000000 +0100
+++ gsasl-1.8.1/lib/m4/gss-extra.m4	2011-01-08 17:32:13.805903205 +0100
@@ -19,7 +19,6 @@
     AC_CHECK_FUNCS([gss_decapsulate_token])
     AC_CHECK_FUNCS([gss_oid_equal])
     AC_CHECK_FUNCS([gss_inquire_mech_for_saslname])
-    AC_CHECK_FUNCS([GSS_C_NT_HOSTBASED_SERVICE])
     if test "$gssapi_impl" != "gss"; then
       AC_CHECK_HEADERS([gssapi.h gssapi/gssapi.h])
       if test "$ac_cv_header_gssapi_h$ac_cv_header_gssapi_gssapi_h" = "nono"; then
@@ -27,6 +26,14 @@
         AC_MSG_WARN([Cannot find gssapi.h or gssapi/gssapi.h, disabling GSSAPI])
       fi
     fi
+    AC_LINK_IFELSE([AC_LANG_PROGRAM([
+#ifdef HAVE_GSSAPI_H
+#include <gssapi.h>
+#else
+#include <gssapi/gssapi.h>
+#endif
+], [gss_OID x = GSS_C_NT_HOSTBASED_SERVICE])],
+[AC_DEFINE([HAVE_GSS_C_NT_HOSTBASED_SERVICE], [1], [Have GSS_C_NT_HOSTBASED_SERVICE symbol])])
     CPPFLAGS="$save_CPPFLAGS"
     LIBS="$save_LIBS"
   fi
